src/PrpShop/PRP/Surface/QMipmap.cpp

```diff
@@ -126,45 +120,42 @@ void QTextureBox::paintEvent(QPaintEvent* evt)
 
 /* QMipmap_Preview */
 QMipmap_Preview::QMipmap_Preview(plCreatable* pCre, QWidget* parent)
-    : QCreatable(pCre, kPreviewMipmap, parent), fLevel()
+    : QCreatable(pCre, kPreviewMipmap, parent)
 {
     plMipmap* tex = plMipmap::Convert(fCreatable);
 
-    QScrollArea* scroll = new QScrollArea(this);
-    fTexture = new QTextureBox(scroll);
-    scroll->setWidget(fTexture);
-
-    QWidget* levelWidget = new QWidget(this);
-    QGridLayout* levelLayout = new QGridLayout(levelWidget);
-    levelLayout->setContentsMargins(4, 4, 4, 4);
-    levelLayout->setHorizontalSpacing(8);
-    QSpinBox* levelSel = new QSpinBox(levelWidget);
-    if (tex->getCompressionType() == plBitmap::kJPEGCompression)
-        levelSel->setRange(0, 0);
-    else
-        levelSel->setRange(0, tex->getNumLevels() - 1);
-    connect(levelSel, QOverload<int>::of(&QSpinBox::valueChanged),
-            this, &QMipmap_Preview::setLevel);
-    levelLayout->addWidget(new QLabel(tr("Level:"), levelWidget), 0, 0);
-    levelLayout->addWidget(levelSel, 0, 1);
-    levelSel->setSizePolicy(QSizePolicy::MinimumExpanding, QSizePolicy::Minimum);
-    QLinkLabel* saveAsLink = new QLinkLabel(tr("Save As..."), levelWidget);
-    levelLayout->addWidget(saveAsLink, 1, 0, 1, 2);
-    connect(fTexture, &QTextureBox::textureChanged, saveAsLink, &QWidget::setEnabled);
-    connect(saveAsLink, &QLinkLabel::activated, fTexture, &QTextureBox::saveAs);
-    fTexture->setTexture(tex);
+    auto levelTabs = new QTabWidget(this);
+    for (size_t level = 0; level < tex->getNumLevels(); level++) {
```

Comment:
Plasma has a weird thing where it doesn't actually store data for any levels where the width or height is 2px or smaller.

We can finally account for that here in PrpShop to hide tabs with invalid levels:
```cpp
if ((tex->getLevelWidth(level) | tex->getLevelHeight(level)) & 0x03)
    continue;
```

Reply:
Huh, good to know. Is this always the case? I'm trying to make sense of [the relevant Plasma code](https://github.com/H-uru/Plasma/blob/da1a02bd09e40532393a561e7c7a1e5cd3e79168/Sources/Plasma/PubUtilLib/plGImage/plMipmap.cpp#L640-L645) and it seems to be specific to DXT compression. It also reads like small sizes would have *some* data, just not in the usual format? :/

Reply:
It definitely might be specific to DXT compression. I know that OpenGL rejected those smaller levels as invalid data when I was implementing texture loading for plGLPipeline, and there were enough places in Plasma doing tests against `& 0x03` to make it apparent that something weird was going on with those.

We also had to workaround this in korman because of similar issues: https://github.com/H-uru/korman/blob/106ef015480fb3dc0ba3cef90491af042458d58b/korlib/texture.cpp#L80-L91

Reply:
Going off of memeories of yester-year, the issue I most remember running into is Cyan's level size calculation for DXT5 compression being wrong. IIRC for a 2x2 mip level, it only allocates a single color block but no alpha block.

Reply:
Did some quick research (read: punched "DXT" into Wikipedia). Apparently, DXT operates on blocks of 4x4 pixels, so it makes sense that it can't work for textures that are smaller than that along either axis.

I still can't really follow what the Plasma code is doing in that case. But based on some experimentation and staring at hex dumps of a couple of mipmaps, I think any levels narrower than 4 pixels are simply stored uncompressed.

This quick and dirty patch gives me reasonable results for e. g. Ahnonay's mipmaps. (Well, as reasonable as you can get for a texture that's 1 or 2 pixels wide, anyway. I agree with [Deledrius' take](https://github.com/H-uru/korman/blob/106ef015480fb3dc0ba3cef90491af042458d58b/korlib/texture.cpp#L87-L89) in the comment that Hoikas linked.)

```patch
--- a/src/PrpShop/PRP/Surface/QMipmap.cpp
+++ b/src/PrpShop/PRP/Surface/QMipmap.cpp
@@ -87,7 +87,11 @@
             dp++;
         }
     }
-    fImage = new QImage(fImageData, tex->getLevelWidth(level),
+    const unsigned char* theActualImageData = fImageData;
+    if (tex->getCompressionType() == plMipmap::kDirectXCompression && (tex->getLevelWidth(level) < 4 || tex->getLevelHeight(level) < 4)) {
+        theActualImageData = static_cast<const unsigned char*>(tex->getLevelData(level));
+    }
+    fImage = new QImage(theActualImageData, tex->getLevelWidth(level),
                         tex->getLevelHeight(level), QImage::Format_ARGB32);
     resize(tex->getLevelWidth(level), tex->getLevelHeight(level));
     update();
```

Reply:
I think it would be better to handle this case correctly in libHSPlasma. AFAICT, the mipmap levels in question *do* have image data stored, so we shouldn't hide them entirely. The special handling for the different image data format should go into `DecompressImage` and not here.